supermarq-benchmarks/supermarq/qcvv/cb.py

```diff
@@ -0,0 +1,800 @@
+
+        fidelities_delta *= fidelities / abs(depths[0] - depths[1])
+        fidelities.columns = pd.MultiIndex.from_product([["fidelity"], fidelities.columns])
+        fidelities_delta.columns = pd.MultiIndex.from_product([["delta"], fidelities_delta.columns])
+
+        if self.experiment._undressed_process:
+            fidelities_delta.loc[:, ("delta", "undressed")] = np.sqrt(
+                ((fidelities_delta.delta / fidelities.fidelity) ** 2).sum(axis=1).values
+            )
+            fidelities.loc[:, ("fidelity", "undressed")] = (
+                fidelities.fidelity.dressed / fidelities.fidelity.undressed
+            )
+            fidelities_delta.loc[:, ("delta", "undressed")] *= fidelities.fidelity.undressed
+
+            for d in depths:
+                expectations.loc[:, ("expectation_mean", d, "undressed")] = (
+                    expectations.expectation_mean[d, "dressed"]
+                    / expectations.expectation_mean[d, "undressed"]
+                )
+                expectations.loc[:, ("expectation_delta", d, "undressed")] = np.sqrt(
+                    (
+                        expectations.expectation_delta[d, "dressed"]
+                        / expectations.expectation_mean[d, "dressed"]
+                    )
+                    ** 2
+                    + (
+                        expectations.expectation_delta[d, "undressed"]
+                        / expectations.expectation_mean[d, "undressed"]
+                    )
+                    ** 2
+                )
+
+        fidelities = pd.concat([fidelities, fidelities_delta], axis=1)
+        fidelities = fidelities.stack(level=[1], future_stack=True).reset_index()
+        expectations = expectations.stack(level=[1, 2], future_stack=True).reset_index()
+
+        self._channel_fidelities = fidelities
+
+        self._channel_expectations: pd.DataFrame = expectations
+
+        self._process_fidelity_estimate = fidelities[
+            fidelities.process == "dressed"
+        ].fidelity.mean()
+
+        self._process_fidelity_estimate_std = np.sqrt(
+            (fidelities[fidelities.process == "dressed"].delta ** 2).sum()
+        ) / len(fidelities)
+
+        if self.experiment._undressed_process:
+            self._undressed_process_fidelity_estimate = fidelities[
+                fidelities.process == "undressed"
+            ].fidelity.mean()
+            self._undressed_process_fidelity_estimate_std = np.sqrt(
+                (fidelities[fidelities.process == "undressed"].delta ** 2).sum()
+            ) / len(fidelities)
+
+    def _process_probabilities(self) -> pd.DataFrame:
+        """Processes the probabilities generated by sampling the circuits into the data structures
+        needed for analyzing the results.
+
+        Returns:
+            A data frame of the full results needed to analyse the experiment.
+        """
+        records = []
+        data: pd.DataFrame = self.data
+        for _, entry in data.iterrows():
```

Comment:
Not for this PR, but maybe we should have a separate issue about optimizing this sort of thing (thinking of the pandas iterrows meme: https://ryxcommar.com/2020/01/15/for-the-love-of-god-stop-using-iterrows/).

But also, I think this is fine because we're not running very large experiments anytime soon, and "Premature optimization is the root of all evil."